scripts/utils/generate_vif/generate_vif.py

```diff
@@ -0,0 +1,264 @@
+        max_voltage = (pdo_value >> 20) & 0x3ff
+        max_voltage_mv = max_voltage * 50
+        min_voltage = (pdo_value >> 10) & 0x3ff
+        min_voltage_mv = min_voltage * 50
+        current = pdo_value & 0x3ff
+        current_ma = current * 10
+        power_mw = (current_ma * max_voltage_mv) // 1000
+        add_xml_element(xml_mem, constants.SINK_PDO_MIN_VOLTAGE,
+                        f'{min_voltage_mv} mV',
+                        {constants.VALUE: str(min_voltage)})
+        add_xml_element(xml_mem, constants.SINK_PDO_MAX_VOLTAGE,
+                        f'{max_voltage_mv} mV',
+                        {constants.VALUE: str(max_voltage)})
+        add_xml_element(xml_mem, constants.SINK_PDO_OP_CURRENT,
+                        f'{current_ma} mA',
+                        {constants.VALUE: str(current)})
+    elif pdo_type == 3:
+        pps = (pdo_value >> 28) & 3
+        if pps:
+            raise ValueError(f'ERROR: Invalid PDO_TYPE {pdo_value}')
+        pps_max_voltage = (pdo_value >> 17) & 0xff
+        pps_max_voltage_mv = pps_max_voltage * 100
+        pps_min_voltage = (pdo_value >> 8) & 0xff
+        pps_min_voltage_mv = pps_min_voltage * 100
+        pps_current = pdo_value & 0x7f
+        pps_current_ma = pps_current * 50
+        power_mw = (pps_current_ma * pps_max_voltage_mv) // 1000
+        add_xml_element(xml_mem, constants.SINK_PDO_MIN_VOLTAGE,
+                        f'{pps_min_voltage_mv} mV',
+                        {constants.VALUE: str(pps_min_voltage)})
+        add_xml_element(xml_mem, constants.SINK_PDO_MAX_VOLTAGE,
+                        f'{pps_max_voltage_mv} mV',
+                        {constants.VALUE: str(pps_max_voltage)})
+        add_xml_element(xml_mem, constants.SINK_PDO_OP_CURRENT,
+                        f'{pps_current_ma} mA',
+                        {constants.VALUE: str(pps_current)})
+    else:
+        raise ValueError(f'ERROR: Invalid PDO_TYPE {pdo_value}')
+    return power_mw
+
+
+def parse_controller_and_data(xml_mem, cad):
+    xml_mem = add_xml_element(xml_mem, cad.basename)
+    for name in cad.data:
+        add_xml_element(xml_mem, name, str(cad.data[name]))
+    parse_node(xml_mem, cad.controller)
+
+
+def parse_array(xml_mem, prop):
+    for member in prop.val:
+        if is_simple_data(member):
+            add_xml_element(xml_mem, prop.name, str(member))
+        elif isinstance(member, list):
+            xml_mem_1 = add_xml_element(xml_mem, prop.name)
+            parse_array(xml_mem_1, member)
+        elif isinstance(member, edtlib.Node):
+            xml_mem_1 = add_xml_element(xml_mem, prop.name)
+            parse_node(xml_mem_1, member)
+        elif isinstance(member, edtlib.ControllerAndData):
+            xml_mem_1 = add_xml_element(xml_mem, prop.name)
+            parse_controller_and_data(xml_mem_1, member)
+        else:
+            ValueError(
+                f'Noticed undefined type : {str(type(member))}, with value {str(member)}')
```

Comment:
Are the `str` casting needed?

Reply:
yes, for storing in xml format otherwise it's throwing errors w.r.to incompatible type